lib/practice_exercises.rb

```diff
@@ -1,13 +1,78 @@
+# frozen_string_literal: true
 
-# Time Complexity: ?
-# Space Complexity: ?
+# Pauline Chane (@PaulineChane on GitHub)
+# Ada Developers Academy C14
+# Data Structures - practice_exercises.rb
+# 09/29/2020
+
+# Time Complexity: O(n). In the worst case scenario for a long string of special characters, the while loop will move
+#                  over one character per loop until the loop is exited, the two lhs/rhs indices meet, or the two
+#                  rhs/lhs indices crossover. In this scenario, there will be string.length loops maximum. With the only
+#                  other operation, string.length, being a non-nested additive O(N) operation that depends on the length
+#                  of the string, the time complexity is about 2N, or O(N) for short.
+# Space Complexity: O(1). Only three variables, valid_char, lhs, and rhs are created, none of which are required to
+#                   expand their memory space depending on the length of the string, constantly occupying the same
+#                   amount of memory. This means space complexity is constant.
 def is_palindrome(string)
-  raise NotImplementedError, "Not implemented yet"
+  valid_char = "abcdefghijklmnopqrstuvwxyz0123456789"
+  # FIX FOR PUNCTUATION
+  lhs = 0
+  rhs = string.length - 1 # O(N) operation for length
+
+  while lhs < rhs do # O(N) operation because max. loops ~= length of string
+    # break into next loop if rhs or lhs are not a valid char
+    if !valid_char.include?(string[lhs].downcase)
+      lhs += 1
+    elsif !valid_char.include?(string[rhs].downcase)
+      rhs -= 1
```

Comment:
Remember that `.include` goes through the string and so has a time complexity of it's own.  

How does that impact your Big-O time complexity?

Reply:
Thank you!! And that's a good point -- my original thought was that the string that include searches, valid_char, will be the same length regardless of the string that is input as it is only searching a single character. For a shorter string this would seem more like O(n^2) or O(nm) (<-- as learned through the other comment) relative to valid_char, but for a very, very long string I would assume this would shrink back to O(n)?